hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/token/TransferTokenTest.java

```diff
@@ -0,0 +1,174 @@
+import com.hedera.services.bdd.junit.HapiTestLifecycle;
+import com.hedera.services.bdd.junit.support.TestLifecycle;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.math.BigInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+import org.apache.tuweni.bytes.Bytes;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.MethodOrderer.OrderAnnotation;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.TestMethodOrder;
+
+@Tag(SMART_CONTRACT)
+@DisplayName("transferToken")
+@SuppressWarnings("java:S1192")
+@HapiTestLifecycle
+@TestMethodOrder(OrderAnnotation.class)
+public class TransferTokenTest {
+
+    private static final String TOKEN_TRANSFER_CONTRACT = "TokenTransferContract";
+    private static final String TOKEN_RECEIVER_CONTRACT = "NestedHTSTransferrer";
+    private static final String FUNGIBLE_TOKEN = "fungibleToken";
+    private static final String ACCOUNT = "account";
+    private static final AtomicReference<Address> fungibleTokenNum = new AtomicReference<>();
+    private static final AtomicReference<Long> transferContractId = new AtomicReference<>();
+    private static final AtomicReference<Long> receiverContractId = new AtomicReference<>();
+
+    @BeforeAll
+    static void beforeAll(@NonNull final TestLifecycle testLifecycle) {
+        testLifecycle.doAdhoc(
+                // Token owner contract that trys to transfer owned tokens
+                uploadInitCode(TOKEN_TRANSFER_CONTRACT),
+                contractCreate(TOKEN_TRANSFER_CONTRACT).gas(1_000_000L).exposingNumTo(transferContractId::set),
+
+                // Contract that is the recipient of the tokens
+                uploadInitCode(TOKEN_RECEIVER_CONTRACT),
+                contractCreate(TOKEN_RECEIVER_CONTRACT).gas(1_000_000L).exposingNumTo(receiverContractId::set),
+
+                // Create the treasury account
+                cryptoCreate(ACCOUNT).balance(100 * ONE_HUNDRED_HBARS),
+
+                // Create fungible token
+                tokenCreate(FUNGIBLE_TOKEN)
+                        .treasury(ACCOUNT)
+                        .initialSupply(100)
+                        .exposingAddressTo(fungibleTokenNum::set),
+
+                // Do associations
+                tokenAssociate(TOKEN_TRANSFER_CONTRACT, FUNGIBLE_TOKEN),
+                tokenAssociate(TOKEN_RECEIVER_CONTRACT, FUNGIBLE_TOKEN),
+
+                // Transfer fungible tokens to the TokenTransferContract contract
+                cryptoTransfer(moving(20, FUNGIBLE_TOKEN).between(ACCOUNT, TOKEN_TRANSFER_CONTRACT)));
+    }
+
+    @Nested
+    @DisplayName("successful when")
+    @Order(1)
+    class SuccessfulTransferTokenTest {
+        @HapiTest
```

Comment:
Would this test be easier to read if some of this syntax noise was abstracted away with useful names?  E.g.,

* give a name to `asHeadlongAddress(Bytes.ofUnsignedLong(transferContractId.get()).toArray())` - perhaps `transferContractIdAsAddress` (or even the shorter `transferContractAddr`) and then that name was used as the argument to `contractCall(...)`?  (Same for `receiverContractId` too of course.)
* Similar for `fungibleTokenNum.get()`?  Might not be worth if you think of it as saving 6 characters but `.get()` is just noise, maybe it reads better without it (even though you have to invent a name).

Though it is effort to invent the names (I know!) and each such name requires a field declaration line, the result would, I think, be that the actual _test_ which you're trying to understand would "pop out" at you (since it would be much shorter - # of lines - and much less crufty) and then would be more comprehensible.

In the future - suppose we had Solidity in the build for these two contracts, perhaps we could then generate Java wrappers for each contract we have!  So you could write `TOKEN_TRANSFER_CONTRACT.approvePublic(fungibleToken, transferContractAddr, 2L).gas(gas_1M)` instead of the 8 lines it takes now?  And it would be syntax checked too!)